homeassistant/components/cookidoo/coordinator.py

```diff
@@ -0,0 +1,71 @@
+"""DataUpdateCoordinator for the Cookidoo integration."""
+
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+
+from cookidoo_api import (
+    Cookidoo,
+    CookidooAuthException,
+    CookidooException,
+    CookidooItem,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_EMAIL
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN, TODO_ADDITIONAL_ITEMS, TODO_INGREDIENTS
+
+_LOGGER = logging.getLogger(__name__)
+
+type CookidooConfigEntry = ConfigEntry[CookidooDataUpdateCoordinator]
+
+
+class CookidooDataUpdateCoordinator(
+    DataUpdateCoordinator[dict[str, list[CookidooItem]]]
+):
+    """A Cookidoo Data Update Coordinator."""
+
+    config_entry: CookidooConfigEntry
+
+    def __init__(self, hass: HomeAssistant, cookidoo: Cookidoo) -> None:
+        """Initialize the Cookidoo data coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=90),
+        )
+        self.cookidoo = cookidoo
+
+    async def _async_update_data(self) -> dict[str, list[CookidooItem]]:
+        try:
+            ingredients = await self.cookidoo.get_ingredients()
+            additional_items = await self.cookidoo.get_additional_items()
+        except CookidooAuthException as e:
+            try:
+                await self.cookidoo.refresh_token()
+            except CookidooAuthException as exc:
+                raise ConfigEntryAuthFailed(
+                    translation_domain=DOMAIN,
+                    translation_key="setup_authentication_exception",
+                    translation_placeholders={
+                        CONF_EMAIL: self.config_entry.data[CONF_EMAIL]
+                    },
+                ) from exc
+            raise UpdateFailed(
+                "Authentication failed but re-authentication was successful, trying again later"
```

Comment:
since you're already translating everything: those can be translated now as well

Reply:
@miaucl I would also suggest not to raise `UpdateFailed` here, instead just write _LOGGER.debug that the token refresh was successful and then return self.data to skip the update

Reply:
it's very new, so you'll most likely just need a rebase. But I'm with @tr4nt0r on this one, no need to raise an `UpdateFailed` here (but you can translate the one below)

Reply:
yes, that is what I am doing. Rebase it is then :)